Date Field Implementation (#804)
deephaven/deephaven-plugins: plugins/ui/docs/components/date_field.md

The `granularity` prop allows you to control the smallest unit that is displayed by a date field . By default, `LocalDate` values are displayed with "DAY" granularity (year, month, and day), and `ZonedDateTime` and `Instant` values are displayed with "SECOND" granularity.

In addition, when a value with a time is provided but you wish to display only the date, you can set the granularity to "DAY". This has no effect on the actual value (it still has a time component), only on what fields are displayed. In the following example, two date fields are synchronized with the same value but display different granularities.

```python
from deephaven import ui

@ui.component
def granularity_example():
    value, set_value = ui.use_state("2021-04-07T18:45:22 UTC")
    return ui.flex(
        ui.date_field(
            label="Date field and time field",
            granularity="SECOND",
            value=value,
            on_change=set_value,
        ),
        ui.date_field(
            label="Date field", granularity="DAY", value=value, on_change=set_value
        ),
        gap="size-150",
        wrap=True,
    )

my_granularity_example = granularity_example()
```

## HTML forms

> Date field supports the `name` prop for integration with HTML forms. The values will be submitted to the server as `ISO 8601` formatted strings according to the granularity of the value. For example, if the date field allows selecting only dates, then strings such as "2023-02-03" will be submitted, and if it allows selecting times, then strings such as "2023-02-03T08:45:00"

Reviewer: ```suggestion
Date field supports the `name` prop for integration with HTML forms. The values will be submitted to the server as `ISO 8601` formatted strings according to the granularity of the value. For example, if the date field allows selecting only dates, then strings such as "2023-02-03" will be submitted, and if it allows selecting times, then strings such as "2023-02-03T08:45:00".
```

```python
from deephaven import ui

my_date_field_forms = ui.form(
    ui.date_field(label="Birth date", name="birthday"),
    ui.button("Submit", type="submit"),
    on_submit=print,
)
```

## Labeling

A visual label should be provided for the date field using the `label` prop. If the date field is required, the `is_required` and `necessity_indicator` props can be used to show a required state.

```python
from deephaven import ui

my_date_field_labeling = ui.flex(
    ui.date_field(label="Date field"),
    ui.date_field(label="Date field", is_required=True, necessity_indicator="icon"),
    ui.date_field(label="Date field", is_required=True, necessity_indicator="label"),
    ui.date_field(label="Date field", necessity_indicator="label"),
)
```

## Events

> Date field s support selection through mouse, keyboard, and touch inputs via the `on_change` prop, which receives the value as an argument.

Reviewer: ```suggestion
Date fields support selection through mouse, keyboard, and touch inputs via the `on_change` prop, which receives the value as an argument.
```

@ui.component
def date_field_validation_behaviour_example():
    return ui.form(
        ui.date_field(
            validation_behavior="native",
            is_required=True,
        )
    )

my_date_field_validation_behaviour_example = date_field_validation_behaviour_example()
```

## Minimum and maximum values

The `min_value` and `max_value` props can also be used to ensure the value is within a specific field. Date field also validates that the end date is after the start date.

```python
from deephaven import ui

my_date_field_basic = ui.date_field(
    label="Date field",
    min_value="2024-01-01",
    default_value="2022-02-03",
)
```

## Label position

By default, the position of a date field's label is above the date field , but it can be moved to the side using the `label_position` prop.

```python
from deephaven import ui

@ui.component
def date_field_label_position_examples():
    return [
        ui.date_field(
            label="Test Label",
        ),
        ui.date_field(
            label="Test Label",
            label_position="side",
        ),
    ]

my_date_field_label_position_examples = date_field_label_position_examples()
```

## Quiet state

> The `is_quiet` prop makes a date field "quiet". This can be useful when the and its corresponding styling should not distract users from surrounding content.

Reviewer: ```suggestion
The `is_quiet` prop makes a date field "quiet". This can be useful when its corresponding styling should not distract users from surrounding content.
```

```python
from deephaven import ui

my_date_field_is_quiet_example = ui.date_field(
    is_quiet=True,
)
```

## Disabled state

> The `is_disabled` prop disables a to prevent user interaction. This is useful when the date field should be visible but not available for selection.

Reviewer: ```suggestion
The `is_disabled` prop disables the date field to prevent user interaction. This is useful when the date field should be visible but not available for selection.
```

The `validation_state` prop can be used to set whether the current date field state is `valid` or `invalid`.

```python
from deephaven import ui

@ui.component
def date_field_help_text_examples():
    return [
        ui.date_field(
            label="Sample Label",
            description="Enter a date field.",
        ),
        ui.date_field(
            label="Sample Label",
            validation_state="valid",
            error_message="Sample invalid error message.",
        ),
        ui.date_field(
            label="Sample Label",
            validation_state="invalid",
            error_message="Sample invalid error message.",
        ),
    ]

my_date_field_help_text_examples = date_field_help_text_examples()
```

## Contextual help

Using the `contextual_help` prop, a `ui.contextual_help` can be placed next to the label to provide additional information about the date field .

```python
from deephaven import ui

date_field_contextual_help_example = ui.date_field(
    label="Sample Label",
    contextual_help=ui.contextual_help(ui.heading("Content tips")),
)
```

## Custom width

> The `width` prop adjusts the width of a date field , and the `max_width` prop enforces a maximum width.

Reviewer: ```suggestion
The `width` prop adjusts the width of a date field, and the `max_width` prop enforces a maximum width.
```

```python
from deephaven import ui

@ui.component
def date_field_width_examples():
    return [
        ui.date_field(
            width="size-3600",
        ),
        ui.date_field(
            width="size-3600",
            max_width="100%",
        ),
    ]

my_date_field_width_examples = date_field_width_examples()
```

## Hide time zone

The time zone can be hidden using the `hide_time_zone` option.

```python
from deephaven import ui

my_hide_time_zone_example = ui.date_field(
    label="Date field",
    default_value="2022-11-07T00:45 America/Los_Angeles",
    hide_time_zone=True,
)
```

## Hour cycle

By default, date field displays times in either a `12` or `24` hour format depending on the user's locale. However, this can be overridden using the `hour_cycle` prop.

```python
from deephaven import ui

date_field_hour_cycle_example = ui.date_field(label="Date field", hour_cycle=24)
```

## Time table filtering

> Date field s can be used to filter tables with time columns.

Reviewer: ```suggestion
Date fields can be used to filter tables with time columns.
```
